### ISSUE TITLE
为什么yolo_training.py中，求pred_box与target的bbox_iou矩阵？

### ISSUE THREAD

**Issue by Bilibilee** [bubbliiiing/yolov4-pytorch#221]
在yolo_training.py中的get_ignore函数里，为什么是求**pred_box**与target的bbox_iou矩阵，再以此作为负样本判断的依据？
我认为应该是求anchors与target的bbox_iou矩阵，yolo_training.py中的get_target函数中也是anchors与target的bbox_iou矩阵，faster rcnn也是求anchors与target的bbox_iou矩阵。

**Comment by bubbliiiing** [comment]
yolo系列一直都是用pred_box和target求iou的，预测的框比较准的就认为要忽略了

**Comment by bubbliiiing** [comment]
faster rcnn 用的和retinanet用的那一系列的思想。和YOLO不同

**Comment by Bilibilee** [comment]
！首先感谢作者大大，这么晚还回复我的问题
我是这么认为的，当yolo模型刚初始化时，随机生成好模型参数。以此开始训练，此时模型的输出pred_box也是随机的，那么拿这个时候的pred_box与target求iou，就无意义了。
所以我认为faster rcnn等算法采用anchor与target求iou
在其他项目中，如yolov3，我也有看到拿anchor与target求iou，代码如下：
`
def build_targets_thres(target, anchor_wh, nA, nC, nGh, nGw):
    ID_THRESH = 0.5
    FG_THRESH = 0.5
    BG_THRESH = 0.4
    nB = len(target)  # number of images in batch
    assert(len(anchor_wh)==nA)

    tbox = torch.zeros(nB, nA, nGh, nGw, 4).cuda()  # batch size, anchors, grid size
    tconf = torch.LongTensor(nB, nA, nGh, nGw).fill_(0).cuda()
    tid = torch.LongTensor(nB, nA, nGh, nGw, 1).fill_(-1).cuda() 
    for b in range(nB):
        t = target[b]
        t_id = t[:, 1].clone().long().cuda()
        t = t[:,[0,2,3,4,5]]
        nTb = len(t)  # number of targets
        if nTb == 0:
            continue

        gxy, gwh = t[: , 1:3].clone() , t[:, 3:5].clone()
        gxy[:, 0] = gxy[:, 0] * nGw
        gxy[:, 1] = gxy[:, 1] * nGh
        gwh[:, 0] = gwh[:, 0] * nGw
        gwh[:, 1] = gwh[:, 1] * nGh
        gxy[:, 0] = torch.clamp(gxy[:, 0], min=0, max=nGw -1)
        gxy[:, 1] = torch.clamp(gxy[:, 1], min=0, max=nGh -1)

        gt_boxes = torch.cat([gxy, gwh], dim=1)                                            
        
        anchor_mesh = generate_anchor(nGh, nGw, anchor_wh)
        anchor_list = anchor_mesh.permute(0,2,3,1).contiguous().view(-1, 4)             
        #print(anchor_list.shape, gt_boxes.shape)
        iou_pdist = bbox_iou(anchor_list, gt_boxes)                                     
        iou_max, max_gt_index = torch.max(iou_pdist, dim=1)                           

        iou_map = iou_max.view(nA, nGh, nGw)       
        gt_index_map = max_gt_index.view(nA, nGh, nGw)

        #nms_map = pooling_nms(iou_map, 3)
        
        id_index = iou_map > ID_THRESH
        fg_index = iou_map > FG_THRESH                                                    
        bg_index = iou_map < BG_THRESH 
        ign_index = (iou_map < FG_THRESH) * (iou_map > BG_THRESH)
        tconf[b][fg_index] = 1
        tconf[b][bg_index] = 0
        tconf[b][ign_index] = -1

        gt_index = gt_index_map[fg_index]
        gt_box_list = gt_boxes[gt_index]
        gt_id_list = t_id[gt_index_map[id_index]]
        #print(gt_index.shape, gt_index_map[id_index].shape, gt_boxes.shape)
        if torch.sum(fg_index) > 0:
            tid[b][id_index] =  gt_id_list.unsqueeze(1)
            fg_anchor_list = anchor_list.view(nA, nGh, nGw, 4)[fg_index] 
            delta_target = encode_delta(gt_box_list, fg_anchor_list)
            tbox[b][fg_index] = delta_target
    return tconf, tbox, tid`
时间不早啦，作者大大早点休息，可以下次讨论

**Comment by bubbliiiing** [comment]
嗯嗯，yolo3也有anchor和gt求iou的过程，为了选取正样本
每个gt对应一个anchor，只有一个正样本
在网络不断训练后pred_box会越来越准确，准确的pred_box就不需要继续训练了，这个是我现在在这里用的思想。
你说的这个是OK的，不过YOLO一向不做密集采样，直到YOLOV5和YOLOX才开始增加正样本点，所以和你现在理解的思想有些许不同，当然并不是一定要像我这样做，每个思想都有它的好处，正样本的点的数量我个人觉得也有必要增加
